### PR TITLE
nixos/dolibarr: add package option

### PR DESCRIPTION
**Pull request by schoettl**
## Description of changes

I added a `package` option to the Dolibarr ERP/CRM service in order to use version 18 from unstable instead of version 16 from 23.05.

Backport to 23.05 would be great.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Module updates) Added a release notes entry if the change is significant
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by h7x4** [comment]
Hello @schoettl, welcome to nixpkgs!

This change is looking fine, but unfortunately I think there's gonna be some problems with the part that is adding you to the maintainer list. Traditionally, nixpkgs has avoided adding people to the maintainer list until they are marked as maintainer of a module or a package. This is to avoid having a bunch of maintainers with nothing to maintain, cluttering the list for no reason. I suggest you either mark yourself as a maintainer of this module, or wait with adding yourself to this list until you either create a new package, module, or just want to mark yourself as maintainer of something else.

File: nixos/modules/services/web-apps/dolibarr.nix
Comment (h7x4) [inline nixos/modules/services/web-apps/dolibarr.nix:62]: We have a neat shorthand for this that's a bit underused at the moment, but we're trying to migrate towards using it almost everywhere.

```suggestion
    package = mkPackageOptionMD pkgs "dolibarr" { };
```

**Comment by h7x4** [comment]
You'll also have to rebase the commits to follow the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

An example of a fitting commit message here would be `nixos/dolibarr: add package option`

File: nixos/modules/services/web-apps/dolibarr.nix
Comment (schoettl) [reply nixos/modules/services/web-apps/dolibarr.nix]: Thanks, I switched to it. Much simpler now :) I remember having seen it in the manual but I wasn't sure if it's endorsed.

**Comment by schoettl** [comment]
Thanks @h7x4 for the quick review and tips! I think I fixed everything now.

**Comment by schoettl** [comment]
Does anyone else want to review, or can this be merged?

**Comment by schoettl** [comment]
Can someone merge this please? Or is there anything blocking a merge?

**Review by jian-lin — APPROVED** [review]

**Comment by schoettl** [comment]
Thank you!

**Comment by RaitoBezarius** [comment]
Please do not forget to page accordingly package maintainers, it's not a pleasant feeling to discover PRs merged of things that you are maintainer of (yes I know, OfBorg is stupid, but `git blame` and the package maintainer field is there at least).

**Comment by jian-lin** [comment]
> Please do not forget to page accordingly package maintainers ... `git blame` is there at least

I do not think it is reviewers' duty to ping related (module maintainers, previous PR authors and reviewers) people for **every** PR of a NixOS module.  If you want that for all PRs, the easiest way I can think of is to teach our CI to do that.  If you just want to be notified by PRs you are interested in, you need to find a way to do that on your own.  By the way, if people know existing programs for that, please share it.

That being said, I will ping people related to the module if the PR is not trivial.  If the PR is trivial,  I probably will not do so.  I think adding a `package` option is trivial, accepted (because there are many similar merged PRs and [this][1]) and probably does not break users' systems.

> the package maintainer field is there

I do not consider **package** maintainers related because it is possible that a package maintainer does not use NixOS.

[1]: https://github.com/NixOS/nixpkgs/issues/50476

**Comment by RaitoBezarius** [comment]
> > Please do not forget to page accordingly package maintainers ... `git blame` is there at least
> 
> I do not think it is reviewers' duty to ping related (module maintainers, previous PR authors and reviewers) people for **every** PR of a NixOS module. If you want that for all PRs, the easiest way I can think of is to teach our CI to do that. If you just want to be notified by PRs you are interested in, you need to find a way to do that on your own. By the way, if people know existing programs for that, please share it.

I strongly disagree with the fact this is not the reviewers' duty to ping related people for every PR regarding a NixOS module, except if such people declined to be pinged. Usually, module authors are package authors and those are kept tightly coupled. I spent a lot of time trying to teach our CI and many other pieces of infrastructure to do a lot of things. Alas, all of that takes time.

Usually, `git blame` / `git log` (on this module, this was quite trivial to obtain from the web UI…) is sufficient IMHO.

> That being said, I will ping people related to the module if the PR is not trivial. If the PR is trivial, I probably will not do so. I think adding a `package` option is trivial, accepted (because there are many similar merged PRs and [this](https://github.com/NixOS/nixpkgs/issues/50476)) and probably does not break users' systems.

I would beg to differ, adding an option like `package` looks trivial, but we had many instances of subtle issues because the way to construct the final `package` requires final tweaks and touchups inside the NixOS module and passing naively a naked package can destroy that final processing phase. I would not expect reviewers who are not expert matter in a package (or the module) to know that aspect of the packaging.

> > the package maintainer field is there
> 
> I do not consider **package** maintainers related because it is possible that a package maintainer does not use NixOS.

You could have verified using `git blame` to conclude that both authors were the same: me.

Either way, I hope I made my point clear, I removed myself from maintenance of Dolibarr because of this answer, I don't share this view of maintenance. I would ask you in the future to please avoid merging PRs on packages/modules I am maintainer of without giving me time to review, a lot of maintainers are like me in that regard, FYI.

**Comment by jian-lin** [comment]
> I spent a lot of time trying to teach our CI and many other pieces of infrastructure to do a lot of things. Alas, all of that takes time.

Thanks for all your contributions.  I do not mean to push you to implement those.  My point is that compared to informing **all** potential reviewers of modules you maintain that you want to be pinged for **every** PRs, including trivial ones, teaching CI to do that is more practical. 

> Usually, git blame / git log (on this module, this was quite trivial to obtain from the web UI…) is sufficient IMHO.

I already know that and I do ping related people for **non-trivial** PRs.

I used to ping related people for trivial ones.  However, I feel that some of them are not responsive.  So I seldom do this nowadays.

Anyway, thanks for the advice.

> You could have verified using git blame to conclude that both authors were the same: me.

I think you miss my point.  My point is that checking the module history **alone** is enough to find **all** people related to the module.

> I would ask you in the future to please avoid merging PRs on packages/modules I am maintainer of without giving me time to review

Will try to do that.  Also, please note that this request from you adds an extra burden to my review workflow, which will have negative effect to my contributions.

> a lot of maintainers are like me in that regard

I am willing to also treat their packages and modules like yours.  However, I cannot do that because I do not know who they are.  I would appreciate it if someone, maybe future me, teaches CI to ping related people for all module PRs.

Also, please note that there are other reviewers like me.

> adding an option like package looks trivial, but we had many instances of subtle issues because the way to construct the final package requires final tweaks and touchups inside the NixOS module and passing naively a naked package can destroy that final processing phase

> I strongly disagree with the fact this is not the reviewers' duty to ping related people for every PR regarding a NixOS module, except if such people declined to be pinged.

I still think it is fine for **trivial** module PRs to be reviewed and merged without noticing its maintainers.  You fail to change my opinion because your argument about "tweaks and touchups" does not make sense to me.

I do not understand what those issues are and why this PR causes them.  Could you elaborate that?  After this PR is applied, if the value of the `package` option is default, isn't this module is equivalent to the old one?  And if they are equivalent, all those "tweaks and touchups" you mention do not matter.

I will probably change my opinion if you help me understand your argument about "tweaks and touchups" more.

> I would not expect reviewers who are not expert matter in a package (or the module) to know that aspect of the packaging.

Same as above: reviewers do not need to know that to review an **equivalent** (definition can be found above) PR.

**Comment by RaitoBezarius** [comment]
I'm snipping because I don't have time to engage on all points here, and I will focus on the interesting part.

> > adding an option like package looks trivial, but we had many instances of subtle issues because the way to construct the final package requires final tweaks and touchups inside the NixOS module and passing naively a naked package can destroy that final processing phase
> 
> > I strongly disagree with the fact this is not the reviewers' duty to ping related people for every PR regarding a NixOS module, except if such people declined to be pinged.
> 
> I still think it is fine for **trivial** module PRs to be reviewed and merged without noticing its maintainers. You fail to change my opinion because your argument about "tweaks and touchups" does not make sense to me.
> 
> I do not understand what those issues are and why this PR causes them. Could you elaborate that? After this PR is applied, if the value of the `package` option is default, isn't this module is equivalent to the old one? And if they are equivalent, all those "tweaks and touchups" you mention do not matter.
> 
> I will probably change my opinion if you help me understand your argument about "tweaks and touchups" more.

Here's an example of a non-trivial module with non-trivial package options which require tweaks and touchups because you need to select the right kernel variant accordingly: https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/tasks/filesystems/zfs.nix#L213-L225.

There's many more such packages where there are tweaks and touchups applied to some package, to create a PHP environment, to create a Python environment, to create a software with plugins, etc.

Creating a package option may seem trivial, but this is dependent on the people who know about it to say whether this will not create issues, otherwise, you are just pushing the churn on the maintainer who will have to pick the pieces.

FYI, this is the first time in my years of contributing to nixpkgs that I have to debate over this, I am surprised that we don't share this common understanding, and you are saying that there are potentially other people not sharing it. I guess, either I don't see it because I usually work on areas which obviously require maintainer expertise or there's a change in culture. In either case, I could propose a change to CONTRIBUTING in that sense. But the concept of a maintainer is to maintain something and to be pinged when there's a relevant change on something.

I did the PR to CI a long time ago to support this: https://github.com/NixOS/ofborg/pull/641 — I cannot shoulder the whole CI mechanism to make it pass, unfortunately.

Feel free to reach out to me over private if you prefer to finish this discussion in private, different handles are available in my maintainer entry.

**Comment by jian-lin** [comment]
> I don't have time to engage on all points here

Thanks for your time.

> Here's an example of a non-trivial module with non-trivial package options which require tweaks and touchups because you need to select the right kernel variant accordingly: https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/tasks/filesystems/zfs.nix#L213-L225.

> There's many more such packages where there are tweaks and touchups applied to some package, to create a PHP environment, to create a Python environment, to create a software with plugins, etc.

> Creating a package option may seem trivial, but this is dependent on the people who know about it to say whether this will not create issues, otherwise, you are just pushing the churn on the maintainer who will have to pick the pieces.

I thought you were talking about this specific change to this specific module.  Does those above apply to this specific change?  I do not think so.

To be clear, I merge this without pinging related people because I am sure this PR is equivalent (definition can be found [here][eq-def]) and does not cause breakages.


> To be clear, this mention do matter because I do not know if you literally merged this without any message or nothing to explain for the maintainer what did you look into.

When reviewing this PR on 2023-11-01 (the same day as h7x4 did his review), I think I have done it right.
1. I checked the whole module to ensure that this change is equivalent (definition can be found [here][eq-def]).
2. I also noticed the format of commit message and PR title is wrong, so I changed the PR title.  I did not comment about commit message format because h7x4 already did that.
3. I noticed that `mkPackageOptionMD` should be used but I did not comment because h7x4 already did that.
4. I added the backport label because this PR was valid to be backported and backporting is requested by the author.
5. I subscribed to this PR so that this PR can be merged in time after requested changes were made.

I will state what I have reviewed clearly in the future.  Thanks for pointing out that.

> This is explicitly in the CONTRIBUTING guide.

Thanks for the pointer.  I have read that guide but forgotten this one:

> Ensure that the module maintainers are notified.

Since pinging maintainers are written in the doc, I will do that for trivial PRs as well in the future.  Sorry for not notifying you in this PR.

BTW, I suggest adding yourself to the `maintainers` of this module.

> But the concept of a maintainer is to maintain something and to be pinged when there's a relevant change on something

I agree.  As I said above, I seldom ping related people for trivial PRs nowadays to save my time because some (or many?) maintainers are just not responsive.

> I did the PR to CI a long time ago to support this: https://github.com/NixOS/ofborg/pull/641 — I cannot shoulder the whole CI mechanism to make it pass, unfortunately.

Good to know that there is a PR for this.  I may help in the future.

> Feel free to reach out to me over private if you prefer to finish this discussion in private

Thanks.  However, I prefer openness and transparence.

[eq-def]: https://github.com/NixOS/nixpkgs/pull/264661#issuecomment-1905047677

**Comment by RaitoBezarius** [comment]
Thank you for your understanding!

> BTW, I suggest adding yourself to the maintainers of this module.

I removed myself from the maintenance of Dolibarr as can be seen in the PR you were mentioned in. I invite anyone who care about it to step up for its continued maintenance.

**Pull request merged**
